// blocks/dashboard/src/config.js
Object.defineProperty(exports, "__esModule", {
  value: true,
});

// Customizable Area Start
exports.dashboardGetUrl = "/bx_block_dashboard/dashboard";
exports.dashboarContentType = "application/json";
exports.dashboarApiMethodType = "GET";
exports.dashboardHost = "<calculated when request is sent>";
exports.dashboarUserAgent = "PostmanRuntime/7.26.5";
exports.dashboarAccept = "*/*";
exports.dashboarAcceptEncoding = "gzip, deflate, br";
exports.dashboarConnection = "keep-alive";
exports.dashboartoken = "";
exports.exampleAPiEndPoint = "EXAMPLE_URL";
exports.exampleAPiMethod = "POST";
exports.putAPiMethod = "PUT";
exports.delAPiMethod = "DELETE";

exports.exampleApiContentType = "application/json";
exports.textInputPlaceHolder = "Enter Text";
exports.labelTitleText = "dashboard";
exports.labelBodyText = "dashboard Body";

exports.btnExampleTitle = "CLICK ME";
// Customizable Area End

exports.endPointApiGetNewCollection = "bx_block_catalogue/catalogues";
exports.endPointApiGetCategoryList =
  "bx_block_categories_sub_categories/categories";
exports.endPointApiGetFeaturedProduct =
  "bx_block_catalogue/catalogues?recommended=true";

exports.endPointApiGetIsCartCreated = "bx_block_cart/carts";
exports.endPointApiGetProductDetails = "/bx_block_catalogue/catalogues/";
exports.endPointApiPostCreateCart = "bx_block_cart/carts";
exports.endPointApiPutItemsToCart = "bx_block_cart/carts/";
exports.SuccessfullyItemAdded = "Added To Cart";
exports.endPointApiIsItemInCart =
  "bx_block_cart/user/carts/has_product?catalogue_variant_id=";
exports.endPointApiPostWishlist = "bx_block_wishlist/wishlists";
exports.endPointApiDelWishlist =
  "bx_block_wishlist/wishlists/remove_catalogue/";
exports.endPointApiPutUpdateCartQuantity = "bx_block_cart/carts/";
exports.endPointApiPostReview = "bx_block_catalogue/reviews";
exports.endPointApiGetProductReview =
  "bx_block_catalogue/reviews/product_reviews?id=";
exports.endPointApiUpdateProductReview = "bx_block_catalogue/reviews/";
exports.endPointApiGetUser="bx_block_banner/banners/web_banners_list"
exports.endPointApiPostNotifyMe="bx_block_catalogue/catalogues_variants/"
