// blocks/filteritems/src/config.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.productApiContentType = "application/json";
exports.apiMethodTypeGet = "GET";
exports.apiMethodTypePost = "POST";
exports.DeleteMethodType = "DELETE";
exports.productAPiEndPoint = "catalogue/catalogues";
exports.brandAPiEndPoint = "catalogue/brands";
exports.tagsAPiEndPoint = "catalogue/tags";
exports.filterbrandAPiEndPoint = "filter_items/filtering?q[brand_id]=";
exports.singleCategoryAPiEndPoint = "filter_items/filtering?q[category_id]=";
exports.cataloguesAPiEndPoint = "filter_items/filtering?";
exports.addToWishlistApiEndPoint = "wishlist/wishlists";
exports.categoryAPIEndPoint = "categories/categories/";

exports.getCartApiEndPoint= "cart/carts"
exports.addToCartApiEndPoint = "cart/carts/"
exports.updateQtyEndPoint = "cart/carts/"
exports.cartHasProductAPIEndPoint = "cart/user/carts/has_product"


// Customizable Area Start
exports.NoProductFound = "No Products found";
exports.NoProductFoundTitle = "Try searching with a different terms or browse other categories.";
exports.buttonTitle = "BROWSE PRODUCTS"
exports.btnExampleTitle = "CLICK ME";
// Customizable Area End
