// blocks/filteritems/src/Filteroptions.tsx
import React from "react";
// Customizable Area Start
import {
  Text,
  TouchableOpacity,
  View,
  SafeAreaView,
  Dimensions,
  TextInput,
  FlatList,
  Image
} from "react-native";
// Customizable Area End
const windowWidth = Dimensions.get("window").width;
import MultiSlider from "@ptomasroos/react-native-multi-slider";
import GreenButton from "../../../components/src/GreenButton";
import styles from "./FilterOptionStyle";
import scale, { verticalScale } from "../../../framework/src/utils/Scale";
import TopHeader from "../../../components/src/TopHeader";
import COLOR_CONST from "../../studio-store-ecommerce-theme/src/AppFonts";
import * as IMG_CONST from "../../studio-store-ecommerce-theme/src/ImageConstants";

import CustomErrorModal from "../../../components/src/CustomErrorModal/CustomErrorModal";
import ApplicationLoader from "../../../components/src/AppLoader/AppLoader";

const themeJson = require('../../studio-store-ecommerce-theme/src/theme.json');

import FilteroptionsController, {
  Props
  //configJSON
} from "./FilteroptionsController";

export default class Filteroptions extends FilteroptionsController {
  constructor(props: Props) {
    super(props);
    // Customizable Area Start
    // Customizable Area End
  }

  // Customizable Area Start
  // Customizable Area End

  renderPrice() {
    if (this.state.pricerange == true) {
      return (
        <View style={styles.rangeConatiner}>
          <Text style={styles.SelectPriceTitle}>Select a price range</Text>
          <View style={styles.rangeRow}>
            <Text style={styles.rangeText}>{themeJson.attributes.currency_type} {this.state.rangeLow}</Text>
            <Text style={styles.rangeText}>{themeJson.attributes.currency_type} {this.state.rangeHigh}</Text>
          </View>
          <View style={styles.rangeInnerConatiner}>
            <MultiSlider
              onValuesChange={value => {
                this.setState({ rangeLow: value[0], rangeHigh: value[1] });
              }}
              markerStyle={{
                backgroundColor: themeJson.attributes.primary_color,
                borderColor: COLOR_CONST.white,
                borderWidth: 2,
                width: scale(13),
                height: scale(13),
                marginTop: scale(2.5)
              }}
              trackStyle={{
                backgroundColor: COLOR_CONST.charcoalGrey,
                height: scale(4)
              }}
              selectedStyle={{ backgroundColor: themeJson.attributes.primary_color }}
              sliderLength={windowWidth - 215}
              values={[this.state.rangeLow, this.state.rangeHigh]}
              min={this.state.rangeMin}
              max={this.state.rangeMax}
              enableLabel={false}
              step={1}
            />
          </View>
          <View style={styles.checkBoxContainer}>
         <TouchableOpacity onPress={() => this.onCheckDiscoutTrue()}>
          {this.state.discount ? (
            <Image source={IMG_CONST.CHECKED} style={styles.CheckIocnStyle} />
          ) : (
              <Image source={IMG_CONST.UNCHECKED} style={styles.CheckIocnStyle} />
            )}
        </TouchableOpacity>
          <Text style={styles.DiscountedStyle}> Discounted Items </Text>
         </View>
        </View>
      );
    }
  }

  renderListItemCategory = (item: any) => {
    return (
      <View style={styles.CategoryFilterConatiner}>
        <TouchableOpacity onPress={() => this.onCheckPressCategory(item)}>
          {item.item.active ? (
            <Image source={IMG_CONST.CHECKED} style={styles.CheckIocnStyle} />
          ) : (
            <Image source={IMG_CONST.UNCHECKED} style={styles.CheckIocnStyle} />
          )}
        </TouchableOpacity>
        <Text
          style={[
            styles.labelTitle,
            {
              color: item.active
                ? COLOR_CONST.pastelRed
                : COLOR_CONST.charcoalGrey
            }
          ]}
        >
          {this.state.subCategory ? item.item.name : item.item.attributes.name}
        </Text>
      </View>
    );
  };

  renderCategory() {
    if (this.state.category === true) {
      return (
        <View style={styles.CategoryConatiner}>
          <View style={styles.searchCatgeory}>
            <Image source={IMG_CONST.SEARCH} style={styles.searchImage} />
            <TextInput
              style={styles.TextInputStyle}
              placeholder="Search for a Category(s)"
              underlineColorAndroid="transparent"
              value={this.state.searchText}
              onChangeText={(value) => this.handleInputChange(value)}
            />
          </View>
          <View style={styles.starightLine} />
          <FlatList
            data={this.state.searchedCategoryFilterList}
            extraData={this.state}
            renderItem={this.renderListItemCategory}
          />
        </View>
      );
    }
  }

  renderListItem = (item: any) => {
    return (
      <View style={styles.BrandFilterConatiner}>
        <TouchableOpacity onPress={() => this.onCheckPressBrand(item)}>
          {item.item.active ? (
            <Image source={IMG_CONST.CHECKED} style={styles.CheckIocnStyle} />
          ) : (
            <Image source={IMG_CONST.UNCHECKED} style={styles.CheckIocnStyle} />
          )}
        </TouchableOpacity>
        <Text
          style={[
            styles.labelTitle,
            {
              color: item.active
                ? COLOR_CONST.pastelRed
                : COLOR_CONST.charcoalGrey
            }
          ]}
        >
          {item.item.attributes.name}
        </Text>
      </View>
    );
  };

  renderTagsListItem = (item: any) => {
    return (
      <View style={styles.BrandFilterConatiner}>
        <TouchableOpacity onPress={() => this.onCheckPressTags(item)}>
          {item.item.active ? (
            <Image source={IMG_CONST.CHECKED} style={styles.CheckIocnStyle} />
          ) : (
            <Image source={IMG_CONST.UNCHECKED} style={styles.CheckIocnStyle} />
          )}
        </TouchableOpacity>
        <Text
          style={[
            styles.labelTitle,
            {
              color: item.active
                ? COLOR_CONST.pastelRed
                : COLOR_CONST.charcoalGrey
            }
          ]}
        >
          {item.item.attributes.name}
        </Text>
      </View>
    );
  };

  renderBrand() {
    if (this.state.brand == true) {
      return (
        <View style={styles.BrandConatiner}>
          <View style={styles.searchCatgeory}>
            <Image source={IMG_CONST.SEARCH} style={styles.searchImage} />
            <TextInput
              style={styles.TextInputStyle}
              placeholder="Search for a Brand(s)"
              underlineColorAndroid="transparent"
              value={this.state.brandsText}
              onChangeText={(value) => this.handleInputChangeBrands(value)}
            />
          </View>
          <View style={styles.starightLine} />
          <FlatList
            data={this.state.searchedBrandsFilterList}
            extraData={this.state}
            renderItem={this.renderListItem}
          />
        </View>
      );
    }
  }
  renderTags() {
    if (this.state.tags === true) {
      return (
        <View style={styles.BrandConatiner}>
          <View style={styles.searchCatgeory}>
            <Image source={IMG_CONST.SEARCH} style={styles.searchImage} />
            <TextInput
              style={styles.TextInputStyle}
              placeholder="Search for a Tag(s)"
              underlineColorAndroid="transparent"
              value={this.state.tagsText}
              onChangeText={(value) => this.handleInputChangeTags(value)}
            />
          </View>
          <View style={styles.starightLine} />
          <FlatList
            data={this.state.searchedTagsFilterList}
            extraData={this.state}
            renderItem={this.renderTagsListItem}
          />
        </View>
      );
    }
  }

  render() {
    return (
      //Merge Engine DefaultContainer
      <SafeAreaView style={styles.container}>
        <TopHeader
          headerTitle={"Filter"}
          onPressLeft={() => this.props.navigation.goBack()}
          headerLeftIconName={IMG_CONST.BACK_ICON}
          headerRight={
            <TouchableOpacity style={{ position: 'absolute', top: verticalScale(30),
            right: scale(20), width: scale(50), height: scale(50), justifyContent: 'center', alignItems: 'center'}} onPress={() => this.clearFilterData()}>
              <Text style={styles.clearText}>Clear</Text>
            </TouchableOpacity>
          }
          navigation={this.props.navigation}
          headerTitleStyle={{ marginLeft: scale(100) }}
          headerTitleViewStyle={{ width: scale(150) }}
        />
        <View style={styles.containerFilter}>
          <ApplicationLoader isFetching={this.state.isFetching} />
          <CustomErrorModal
            showModal={this.state.customErrorModal}
            message={this.state.customErrorMessage}
            hideErrorModal={() => this.setState({ customErrorModal: false })}
          />
          <View style={styles.LeftContainerFilter}>
            <TouchableOpacity
              style={{
                width: scale(119.2),
                height: scale(54.6),
                justifyContent: "center",
                backgroundColor: this.changeColor("price")
              }}
              onPress={() => this._onPressChange("price")}
            >
              <Text style={styles.priceRangeTitle}>Price Range</Text>
            </TouchableOpacity>
            <TouchableOpacity
              style={{
                width: scale(119.2),
                height: scale(54.6),
                justifyContent: "center",
                backgroundColor: this.changeColor("brand")
              }}
              onPress={() => this._onPressChange("brand")}
            >
              <Text style={styles.priceRangeTitle}>Brand</Text>
            </TouchableOpacity>
            <TouchableOpacity
              style={{
                width: scale(119.2),
                height: scale(54.6),
                justifyContent: "center",
                backgroundColor: this.changeColor("tags")
              }}
              onPress={() => this._onPressChange("tags")}
            >
              <Text style={styles.priceRangeTitle}>Tags</Text>
            </TouchableOpacity>
            <TouchableOpacity
              style={{
                width: scale(119.2),
                height: scale(54.6),
                justifyContent: "center",
                backgroundColor: this.changeColor("category")
              }}
              onPress={() => this._onPressChange("category")}
            >
              <Text style={styles.priceRangeTitle}>Category</Text>
            </TouchableOpacity>
          </View>
          <View>
            {this.renderPrice()}
            {this.renderBrand()}
            {this.renderTags()}
            {this.renderCategory()}
          </View>
        </View>
        <View style={styles.ButtonConatiner}>
          <View style={styles.InnerConatiner}>
            <TouchableOpacity
              style={styles.buttonCustom}
              onPress={() => this.props.navigation.goBack()}
            >
              <Text style={styles.customTxtStyle}>CANCEL</Text>
            </TouchableOpacity>
            <GreenButton
              title="APPLY FILTER"
              customStyle={styles.buttonCustom}
              customTxtStyle={styles.customTxtStyle}
              onPress={() => this.onPressApplyFilter()}
            />
          </View>
        </View>
      </SafeAreaView>
      //Merge Engine End DefaultContainer
    );
  }
}

// Customizable Area Start

// Customizable Area End
