// blocks/payments/src/ColorConstants.js
export default COLOR_CONST = {
    primaryThemeColor: 'rgb(36,39,36)',
    secondaryThemeColor: 'rgb(212,201,109)',
    primaryThemeGradient: 'rgb(36,39,36)',
    secondaryThemeGradient: 'rgb(36,39,36)',
    tertiaryThemeGradient: 'rgb(36,39,36)',
    greenThemeColor: '#3bc490',
    focusDarkColor: '#3b5898',
    subscribeGreen: '#43c6ac',
    cartGreen: '#3bc48f',
    darkishBlue: '#004491',
    pastelRed: '#e65e52',
    defaultBgColor: '#f2f2f2',
    charcoalGrey: '#3e454f',
    darkGreyBlue: '#31414f',
    greenButton: '#29af9d',
    greenBlue: '#43b7a7',
    borderduckEggBlue: '#eaf0fb',
    paleGrey: '#f6f8fa',
    paleGreyTwo: '#eef2f6',
    paleGreyThree: '#f4f5f6',
    paleGreyFour: '#eff2f5',
    paleGreyFive: '#f6f8fa',
    modalTransparentBg: 'rgba(34,34,34, 0.8)',
    LightlightNavy: '#195aa4',
    boderdarkishBlueTwo: '#30458c',
    lightBlueGrey: '#d8e2ea',
    uploadBG: '#f3f4f4',
    coolGrey: '#8c9096',
    darkishBlueTwo: '#054995',
    coolGreyTwo: "#8b8f95",
    whiteTwo: "#f9f9f9",
    darkBlueGrey: "#101c2f",
    white: '#ffffff',
    lightWhite: '#fafeff',
    buttonColor: '#8ca9b9',
    darkTitle: '#202326',
    appBgColor: '#f5f5f7',
    lightGreyText: '#d8d8d8',
    redAlertColor: '#ff0000',
    darkGreyText: '#616266',
    inActiveDotLightGrey: '#d6d7dc',
    tickColor: '#8ca9b9',
    black: '#000000',
    bellLightGrey: '#f1f3f6',
    readCellBgColor: '#f9fbfd',
    savedCardsGrey: '#9da8ba',
    transparent: 'rgba(0,0,0,0)',
    text: '#212529',
    primary: '#007bff',
    success: '#28a745',
    error: '#dc3545',
    greenyBlue: '#43b7a8',
    thankYouGrey: '#788aa7',
    dollarValue: '#213052',
    whiteThree: '#dbdbdb',
    darkGrey: '#1e1f21',
    lightNotificationColor: '#d8e4f0',
    newNotificationColor: '#becfe0',
    logoGradientFirst: '#d8e4f0',
    logoGradientSecond: '#becfe0',
};

export const FONTS = {
    GTWalsheimProBold: 'GTWalsheimPro-Bold',
    GTWalsheimProLight: 'GTWalsheimPro-Light',
    GTWalsheimProMedium: 'GTWalsheimPro-Medium',
    GTWalsheimProRegular: 'GTWalsheimPro-Regular',
}
