// blocks/payments/src/config.js
Object.defineProperty(exports, "__esModule", {
  value: true
});

// Customizable Area Start
exports.validationApiContentType = "application/json";
exports.checkoutStatusMethodType = "GET";
exports.GetMethodType = "GET";
exports.checkoutAPiEndPoint = "/payments/hyperpay_payments";
exports.checkoutStatus = "/payments/hyperpay_payments/payment_status?checkout_id=";
exports.hyperpayAPIMethodType = "POST";
exports.exampleApiContentType = "application/json";
exports.textInputPlaceHolder = "Enter Text";
exports.labelTitleText = "payments";
exports.labelBodyText = "payments Body";

exports.btnExampleTitle = "CLICK ME";
exports.httpPostType = "POST"
exports.httpGetType = "GET";
exports.razorpayApiContentType = "application/json";
exports.errorOrderNotValid = "Please enter all fields."
exports.title = "Enter Order Id"
exports.razorpayAPiEndPoint = "payment_razorpay/razorpays";
exports.ordersAPiEndPoint = "order_management/orders";

exports.createRazorpayApiEndPoint = "payment_razorpay/razorpays";
exports.verifyRazorpayApiEndPoint =
  "payment_razorpay/razorpays/varify_signature";

exports.userProfileApiEndPoint = "profile/profile";

exports.cartHasProductAPIEndPoint = "cart/user/carts/has_product";

// Customizable Area End
