// components/src/AppFooter/config.js
Object.defineProperty(exports, "__esModule", {
  value: true,
});

// Customizable Area Start
exports.validationApiContentType = "application/json";
exports.validationApiMethodType = "GET";
exports.exampleAPiMethod = "POST";
exports.exampleApiContentType = "application/json";
exports.btnExampleTitle = "CLICK ME";

/*** Methods List */
exports.apiPostMethod = "POST";
exports.apiPutMethod = "PUT";
exports.apiGetMethod = "GET";
exports.apiDeleteMethod = "DELETE";

/** API for Connected Accounts */
exports.getAllSocialAccountsApiEndPoint =
  "bx_block_social_media_account/social_auths";
exports.postActivateSocialAccountsAPIEndPoint =
  "bx_block_social_media_account/social_auths/connect";
exports.deleteActivateSocialAccountAPiEndPoint =
  "bx_block_social_media_account/social_auths";

exports.baseUrl =
  "https://ecommerceapp-39394-ruby.39394.dev.ap-southeast-1.aws.svc.builder.ai";

exports.getHelpCenterAPIEndPoint = "bx_block_help_center/help_centers";

//brand_settings API
exports.brandSettingAPIEndPoint = 'bx_block_store_profile/brand_settings';


// Customizable Area End
