// blocks/shoppingcart/src/assetsWeb.tsx
export const imgPasswordVisible = require("../assets/ic_password_visible.png");
export const imgPasswordInVisible = require("../assets/ic_password_invisible.png");
export const emptyCart =require("../assets/cart-empty-icn.png");
export const checkCheck = require('../assets/Group 3945.svg');
export const addressImage = require('../assets/address.png');
export const closeImg = require('../assets/Path 10545.svg');
export const emptyCheck = require('../assets/Ellipse 77.svg');
export const closeIcon = require('../assets/Page 1 (1).svg');
export const fbIcon =require('../assets/facebook.png');
export const GoogleIcon = require('../assets/googlePlus.png');
export const backIcon = require("../assets/backIcon.png");
export const notificationIcon = require("../assets/notificationsIcon.png");
export const checkboxSelected = require("../assets/selectedCheckbox.png");
export const checkboxUnSelected = require("../assets/checkboxUnselected.png");
export const blankAddress = require("../assets/blankAddress.png");
export const tickAddress = require("../assets/tickAddress.png");
export const cartEmpty = require("../assets/cartEmpty.png");
export const couponTick = require("../assets/couponTick.png");
