// blocks/shoppingcart/src/config.js
Object.defineProperty(exports, "__esModule", {
  value: true
});

// Customizable Area Start
exports.ApiContentType = "application/json";
exports.validationApiContentType = "application/json";
exports.validationApiMethodType = "GET";
exports.ApiMethodPostType = "POST";
exports.apiMethodTypePost = "POST";
exports.apiMethodTypeGet = "GET"
exports.apiMethodTypePut = "PUT";
exports.apiMethodTypeDel = "DELETE"
exports.exampleAPiEndPoint = "EXAMPLE_URL";
exports.exampleAPiMethod = "POST";
exports.exampleApiContentType = "application/json";
exports.textInputPlaceHolder = "Enter Text";
exports.labelTitleText = "shoppingcart";
exports.labelBodyText = "shoppingcart Body";
exports.ApplyCouponApiEndPoint = "order_management/orders/";
exports.cartListAPiEndPoint = "cart/carts"
exports.updateQtyEndPoint = "cart/carts/"
exports.emptyCartEndPoint = "cart/carts/"
exports.removeCouponEndPoint = "order_management/orders/"
exports.cartHasProductAPIEndPoint = "cart/user/carts/has_product"
exports.addToWishlistApiEndPoint = "wishlist/wishlists";
exports.addressApiEndPoint = "order_management/addresses"



exports.btnExampleTitle = "CLICK ME";
exports.couponCodeErrorMsg = "Oops! Coupon Code Is Not Valid"
// Customizable Area End
