// blocks/dashboard/src/assets.tsx
export const bannerImg = require("../assets/Desktop-2.jpg");
export const pinkLady = require("../assets/pinkLady.png");
export const potraitHand = require("../assets/potraitHand.png");
export const luxuryProduct = require("../assets/luxuryProduct.png");
export const Serviceicon1 = require("../assets/truck-icon.svg");
export const Serviceicon2 = require("../assets/payment.svg");
export const Serviceicon3 = require("../assets/return.svg");
export const Serviceicon4 = require("../assets/support.svg");
export const available = require("../assets/available.png");
export const sold = require("../assets/soldOut.png");

export const CardImage1 = require("../assets/card-1-image.jpg");
export const CardImage2 = require("../assets/card-2-image.jpg");
export const CardImage3 = require("../assets/card-3-image.jpg");

export const OfferCardImage1 = require("../assets/offer-card1.jpg");
export const OfferCardImage2 = require("../assets/offer-card2.jpg");
export const OfferCardImage3 = require("../assets/offer-card3.jpg");
export const heartFav = require("../assets/heart-fav.png");
export const closebutton = require("../assets/closebutton.png");
