// blocks/profilebio/src/config.js
Object.defineProperty(exports, "__esModule", {
  value: true
});


exports.ApiContentType = "application/json";
exports.EditApiContentType = "multipart/form-data"
exports.GetMethodType = "GET";
exports.PostMethodType = "POST";
exports.PutMethodType = "PUT";
exports.DeleteMethodType = "DELETE";
exports.apiMethodTypePut = "PUT"
exports.getWishlistApiEndPoint = "wishlist/wishlists";
exports.cartHasProductEndPoint = "cart/user/carts/has_product"
exports.exampleApiContentType = "application/json";
exports.textInputPlaceHolder = "Enter Text";
exports.labelTitleText = "profilebio";
exports.labelBodyText = "profilebio Body";
exports.changePasswordApiEndPoint = "profile/password";
exports.userProfileApiEndPoint = "profile/profile";
// Customizable Area Start
exports.NoProductFound = "No Products found";
exports.NoProductFoundTitle = "Try searching with a different terms or browse other categories.";
exports.buttonTitle = "BROWSE PRODUCTS"
exports.btnExampleTitle = "CLICK ME";
// Customizable Area End
